pyk/src/tests/unit/test_kcfg.py

```diff
@@ -849,3 +850,69 @@ def test_no_cell_rewrite_to_dots() -> None:
 
     result = no_cell_rewrite_to_dots(term)
     assert result == term
+
+
+CREATE_SPLIT_BY_NODES_TEST_DATA: Final = (
+    (CTerm.bottom(), [CTerm.top(), CTerm.bottom()], None),
+    (CTerm.top(), [CTerm.top(), CTerm.bottom()], None),
+    (
+        CTerm.top(),
+        [CTerm(config('X')), CTerm(config('Y'))],
+        None,  # todo: support split from top, because top means anything can be matched
+    ),
+    (
+        CTerm.top(),
+        [CTerm.top(), CTerm.top()],
+        KCFG.Split(
+            KCFG.Node(1, CTerm.top()), [(KCFG.Node(2, CTerm.top()), CSubst()), (KCFG.Node(3, CTerm.top()), CSubst())]
+        ),
+    ),
+    (
+        CTerm(config('X')),
+        [CTerm(config('X')), CTerm(config('Y')), CTerm(config('Z'))],
+        KCFG.Split(
+            KCFG.Node(1, CTerm(config('X'))),
+            [
+                (KCFG.Node(2, CTerm(config('X'))), CSubst(Subst({'X': KVariable('X')}))),
+                (KCFG.Node(3, CTerm(config('Y'))), CSubst(Subst({'X': KVariable('Y')}))),
+                (KCFG.Node(4, CTerm(config('Z'))), CSubst(Subst({'X': KVariable('Z')}))),
+            ],
+        ),
+    ),
+    (CTerm(config('X')), [CTerm(config('Y')), CTerm(KApply('<bot>', [KVariable('Z')]))], None),
+    (
+        CTerm(config('X'), [ge('X', 0), lt('X', 10)]),
+        [CTerm(config('Y'), [ge('Y', 0)]), CTerm(config('Z'), [ge('Z', 5)])],
+        KCFG.Split(
+            KCFG.Node(1, CTerm(config('X'), [ge('X', 0), lt('X', 10)])),
+            [
+                (KCFG.Node(2, CTerm(config('Y'), [ge('Y', 0)])), CSubst(Subst({'X': KVariable('Y')}), [lt('Y', 10)])),
```

Comment:
Note that this target doesn't meet the implication relationship with source. So the CTerm of target and CSubst.apply(target) are not logically equal. But source -> CSubst.apply(target) can always be true.

Reply:
I think, in that case, we should actually have it fail to build the split.

The semantics of a `split` are that:

- The split targets should be mutually exclusive (no overlap).
- The split targets should be total on the original (the union of the split targets == the split source).

Reply:
It's impossible to check these properties without the Haskell backend and semantics definition. As without them, the functions inside CTerms are uninterpreted. 

We can introduce a tool called `CFGChecker` in the kcfg package for automatic checking. I want the `KCFG` class to be only used for its syntax definition and syntax-level (static-semantics level) manipulation.

Also, I think edge types in kcfg can be reduced to rewrite and implies (or refine). It doesn't seem necessary to distinguish these edge types, except for printing. If necessary, we can identify them at runtime.

Reply:
I want to make it simple and I think it will be acceptable for merging nodes. Because if a mergeable split is correct, the generated split will always be correct.

Reply:
Makes sense.